File: dotnet/Game.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Michitai;
using Michitai.Multiplayer;
using Michitai.Multiplayer.Time;
using Michitai.Multiplayer.Games;
using Michitai.Multiplayer.Players;
using Michitai.Multiplayer.Matchmaking;
using Michitai.Multiplayer.Matchmaking.Requests;
using Michitai.Multiplayer.Rooms;
using Michitai.Multiplayer.Rooms.Actions;
using Michitai.Multiplayer.Rooms.Updates;
using Michitai.Multiplayer.Leaderboard;
using Michitai.Multiplayer.Errors;

public class Game
{
    private static Multiplayer? client;
    private static readonly Dictionary<string, PlayerInfo> players = new();

    public static async Task Main()
    {
        Console.WriteLine("=== MICHITAI Game SDK - ALL THREE DEMOS + TIME + LEADERBOARD ===\n");

        var logger = new ConsoleLogger();
        client = new Multiplayer("YOUR_API_TOKEN", "YOUR_PRIVATE_TOKEN", logger: logger);

        Console.WriteLine("[INIT] SDK initialized successfully\n");

        try
        {
            // Three main demos
            await RunDemoWithJoinByRequests();        // Demo 1: With approval
            await CleanupEverything();

            await RunDemoWithoutJoinByRequests();     // Demo 2: Direct join
            await CleanupEverything();

            await RunDemoDirectRoom();                // Demo 3: Direct room
            await CleanupEverything();

            // Common tests (run once)
            await RunCommonTests();
        }
        catch (Exception ex)
        {
            Console.WriteLine($"[FATAL] Unexpected error: {ex.Message}");
        }

        Console.WriteLine("\n=== All Demos Finished - All Endpoints Covered ===");
    }

    // ====================== COMMON TESTS (Time + Leaderboard + Game Data) ======================
    private static async Task RunCommonTests()
    {
        Console.WriteLine("=== COMMON TESTS: TIME, LEADERBOARD, GAME DATA ===\n");

        // Game Data
        await SafeExecute(async () =>
        {
            var gd = await Games.GetGameData<GameData>(client!);
            Console.WriteLine($"[GAME DATA] Retrieved global data");
            GameData gameData = gd.GameData;
            Console.WriteLine($"[GAME DATA] Event: {gameData.CurrentEvent}, Version: {gameData.Version}");
            await Games.UpdateGameData(client!, new GameData { CurrentEvent = "SpringFestival", Version = "1.2.3" });
            Console.WriteLine("[GAME DATA] Global data updated");
        }, "Game Data");

        // Time API
        await SafeExecute(async () =>
        {
            var time = await Time.GetServerTime(client!);
            Console.WriteLine($"[TIME] Server UTC: {time.Utc}");
        }, "GetServerTime");

        await SafeExecute(async () =>
        {
            var timeOffset = await Time.GetServerTimeWithOffset(client!, 3);
            Console.WriteLine($"[TIME] Server UTC+3: {timeOffset.Utc}");
        }, "GetServerTimeWithOffset");

        // Leaderboard
        await SafeExecute(async () =>
        {
            var lb = await Leaderboard.GetLeaderboardAsync<PlayerData>(client!, new[] { "level", "wins" }, limit: 10);
            Console.WriteLine($"[LEADERBOARD] Top {lb.Leaderboard.Count} players loaded");
            if (lb.Leaderboard.Count > 0)
                Console.WriteLine($"[LEADERBOARD] #1: {lb.Leaderboard[0].Player_name}, Level: {lb.Leaderboard[0].PlayerData.Level}");
        }, "GetLeaderboard");

        Console.WriteLine();
    }

    // ===================================================================
    // DEMO 1: MATCHMAKING WITH JOIN REQUESTS
    // ===================================================================
    private static async Task RunDemoWithJoinByRequests()
    {
        Console.WriteLine("\n=== DEMO 1: MATCHMAKING WITH JOIN REQUESTS ===\n");
        await SetupPlayers();

        string matchmakingId = await CreateMatchmakingLobby("DEMO 1 Matchmaking", joinByRequests: true);

        string req1 = await RequestToJoinMatchmaking(players["p1"].Token, matchmakingId, null);
        await CheckJoinRequestStatus(players["p1"].Token, req1);
        await ApproveJoinRequest(players["host"].Token, req1);

        string req2 = await RequestToJoinMatchmaking(players["p2"].Token, matchmakingId, new PlayerData());
        await CheckJoinRequestStatus(players["p2"].Token, req2);

        await GetCurrentMatchmakingStatus();

        await ApproveJoinRequest(players["host"].Token, req2);

        await GetCurrentMatchmakingStatus();
        await GetMatchmakingPlayers();

        string roomId = await StartMatchmakingAndCreateRoom();
        await RunGameRoomFlow(roomId, isFromMatchmaking: true);
    }

    // ===================================================================
    // DEMO 2: MATCHMAKING DIRECT JOIN
    // ===================================================================
    private static async Task RunDemoWithoutJoinByRequests()
    {
        Console.WriteLine("\n=== DEMO 2: MATCHMAKING DIRECT JOIN ===\n");
        await SetupPlayers();

        string matchmakingId = await CreateMatchmakingLobby("DEMO 2 Matchmaking", joinByRequests: false);

        foreach (var p in players.Values)
        {
            if (players["host"].Token == p.Token)
                continue;

            await JoinMatchmakingDirectly(p.Token, matchmakingId, new PlayerData());
        }

        await GetCurrentMatchmakingStatus();
        await GetMatchmakingPlayers();

        string roomId = await StartMatchmakingAndCreateRoom();
        await RunGameRoomFlow(roomId, isFromMatchmaking: true);
    }

    // ===================================================================
    // DEMO 3: DIRECT ROOM (No Matchmaking)
    // ===================================================================
    private static async Task RunDemoDirectRoom()
    {
        Console.WriteLine("\n=== DEMO 3: DIRECT ROOM CREATION ===\n");
        await SetupPlayers();

        var create = await Rooms.CreateRoomAsync<PlayerData, RulesData>(client!, players["host"].Token, "Direct Battle Arena", 4);
        string roomId = create.Room_id;

        await JoinRoom(players["p1"].Token, roomId);
        await JoinRoom(players["p2"].Token, roomId);

        await RunGameRoomFlow(roomId, isFromMatchmaking: false);
    }

    // ====================== SETUP & CLEANUP ======================
    private static async Task SetupPlayers()
    {
        Console.WriteLine("[SETUP] Registering players...");

        var h = await RegisterPlayer("GameHost", new { level = 15, rank = "gold" });
        var p1 = await RegisterPlayer("PlayerOne", new { level = 12, rank = "silver" });
        var p2 = await RegisterPlayer("PlayerTwo", new { level = 10, rank = "bronze" });

        players["host"] = new PlayerInfo { Id = h.Player_id, Token = h.Private_key, Name = "GameHost" };
        players["p1"] = new PlayerInfo { Id = p1.Player_id, Token = p1.Private_key, Name = "PlayerOne" };
        players["p2"] = new PlayerInfo { Id = p2.Player_id, Token = p2.Private_key, Name = "PlayerTwo" };

        foreach (var p in players.Values)
            await AuthenticatePlayer(p.Token);

        foreach (var p in players.Values)
            await SendPlayerHeartbeat(p.Token);

        await GetAllPlayersList();

        foreach (var p in players.Values)
        {
            var data = await GetPlayerData(p.Token);

            var player = data.PlayerData;

            player.Level++;

            await UpdatePlayerData(p.Token, player);

            data = await GetPlayerData(p.Token);
        }
    }

    private static async Task CleanupEverything()
    {
        Console.WriteLine("\n[CLEANUP] Final cleanup...");

        foreach (var p in players.Values)
        {
            await SafeExecute(async () => await Players.LogoutPlayerAsync(client!, p.Token), $"Logout {p.Name}");
        }

        players.Clear();
    }

    // ====================== HELPER METHODS ======================
    private static async Task<PlayerRegisterResponse> RegisterPlayer(string name, object? data = null)
    {
        var reg = await Players.RegisterPlayer(client!, name, data);
        Console.WriteLine($"[REGISTER] {name} registered");
        return reg;
    }

    private static async Task<PlayerAuthResponse<PlayerData>> AuthenticatePlayer(string token)
    {
        var auth = await Players.AuthenticatePlayer<PlayerData>(client!, token);
        Console.WriteLine($"[AUTH] {auth!.Player!.Player_name} authenticated");
        return auth;
    }

    private static async Task<PlayerHeartbeatResponse> SendPlayerHeartbeat(string token)
    {
        var heartbeat = await Players.SendPlayerHeartbeatAsync(client!, token);
        Console.WriteLine($"[HEARTBEAT] Player heartbeat sent");
        return heartbeat;
    }

    private static async Task GetAllPlayersList()
    {
        var list = await Games.GetAllPlayers(client!);
        Console.WriteLine($"[PLAYERS LIST] Total: {list.Count}");

        foreach (PlayerShort player in list.Players)
        {
            Console.WriteLine($"[PLAYERS LIST] Id: {player.Id}, Name: {player.Player_name}, Online: {player.Is_online}, Login: {player.Last_login}, Created: {player.Created_at}");
        }
    }

    private static async Task<PlayerDataResponse<PlayerData>> GetPlayerData(string token)
    {
        var data = await Players.GetPlayerData<PlayerData>(client!, token);
        Console.WriteLine($"[PLAYER DATA] Player data retrieved");
        return data;
    }

    private static async Task<SuccessResponse> UpdatePlayerData(string token, PlayerData data)
    {
        var res = await Players.UpdatePlayerData(client!, token, data);
        Console.WriteLine($"[PLAYER DATA] Player data updated");
        return res;
    }

    private static async Task<string> CreateMatchmakingLobby(string matchmakingName, bool joinByRequests)
    {
        RulesData rules = new RulesData { Mode = "tdm", Map = "arena" };

        PlayerData playerData = new PlayerData { Level = 3, Rank = "Diamond" };

        var res = await Requests.CreateMatchmakingLobbyAsync<PlayerData, RulesData>(client!, players["host"].Token,
            matchmakingName, 4, false, joinByRequests, false, false, playerData, rules);
        Console.WriteLine($"[MATCHMAKING] Lobby created (requests={joinByRequests})");
        return res.Matchmaking_id;
    }

    private static async Task<string> RequestToJoinMatchmaking(string token, string matchmakingId, PlayerData? playerData = null)
    {
        var req = await Requests.RequestToJoinMatchmakingAsync<PlayerData>(client!, token, matchmakingId, playerData);
        Console.WriteLine($"[REQUEST] Sent: {req.Request_id}");
        return req.Request_id;
    }

    private static async Task CheckJoinRequestStatus(string token, string requestId)
    {
        var status = await Requests.CheckJoinRequestStatusAsync(client!, token, requestId);
        Console.WriteLine($"[REQUEST STATUS] {status.Request.Status}");
    }

    private static async Task ApproveJoinRequest(string hostToken, string requestId)
    {
        var resp = await Requests.RespondToJoinRequestAsync(client!, hostToken, requestId, EMatchmakingRequestAction.Approve);
        Console.WriteLine($"[APPROVE] {resp.Message}");
    }

    private static async Task JoinMatchmakingDirectly(string token, string matchmakingId, PlayerData? playerData = null)
    {
        await Matchmaking.JoinMatchmakingDirectlyAsync<PlayerData>(client!, token, matchmakingId, playerData);
        Console.WriteLine("[JOIN] Player joined directly");
    }

    private static async Task GetCurrentMatchmakingStatus()
    {
        var s = await Matchmaking.GetCurrentMatchmakingStatusAsync<RulesData>(client!, players["host"].Token);
        Console.WriteLine($"[MATCHMAKING STATUS] Players: {s.Matchmaking?.Current_players ?? 0}");
    }

    private static async Task GetMatchmakingPlayers()
    {
        var list = await Matchmaking.GetMatchmakingPlayersAsync<PlayerData>(client!, players["host"].Token);
        Console.WriteLine($"[MATCHMAKING PLAYERS] {list.Players.Count} players");
    }

    private static async Task<string> StartMatchmakingAndCreateRoom()
    {
        var start = await Matchmaking.StartGameFromMatchmakingAsync(client!, players["host"].Token);
        Console.WriteLine($"[START] Room created: {start.Room_id}");
        return start.Room_id;
    }

    private static async Task JoinRoom(string token, string roomId)
    {
        await Rooms.JoinRoomAsync<PlayerData>(client!, token, roomId);
        Console.WriteLine($"[ROOM] Player joined room");
    }

    private static async Task RunGameRoomFlow(string roomId, bool isFromMatchmaking)
    {
        Console.WriteLine("\n=== GAME ROOM FLOW ===\n");

        await Rooms.GetRoomsAsync<RulesData>(client!);
        await Rooms.GetCurrentRoomAsync<RulesData>(client!, players["host"].Token);

        foreach (var p in players.Values)
        {
            await SafeExecute(async () => 
            {
                var req = new SubmitAction<ActionData>(ERoomTargetPlayers.Host, "player_ready", new ActionData { Ready = true });
                await Actions.SubmitActionAsync(client!, p.Token, req);
            }, $"SubmitAction {p.Name}");
        }

        await SafeExecute(async () =>
        {
            var pending = await Actions.GetPendingActionsAsync<ActionData>(client!, players["host"].Token);
            Console.WriteLine($"[PENDING ACTIONS] {pending.Actions.Count} actions");
        }, "GetPendingActions");

        await SafeExecute(async () =>
        {
            var req = new UpdatePlayers<UpdateData>(ERoomTargetPlayers.All, "game_start", new UpdateData { Round = 1, Message = "Game Started!" });
            await Updates.UpdatePlayersAsync<UpdateData>(client!, players["host"].Token, req);
        }, "Send Room Update");

        foreach (var p in players.Values)
            await SafeExecute(async () => await Updates.PollUpdatesAsync<UpdateData>(client!, p.Token, new PollUpdates(ERoomTargetPlayers.Host)), $"PollUpdates {p.Name}");

        await Rooms.GetRoomPlayersAsync<PlayerData>(client!, players["host"].Token);

        foreach (var p in players.Values)
            await SafeExecute(async () => await Rooms.SendRoomHeartbeatAsync(client!, p.Token), $"RoomHeartbeat {p.Name}");

        await SafeExecute(async () => await Rooms.LeaveRoomAsync(client!, players["host"].Token), $"LeaveRoom {players["host"].Name}");
    }

    private static async Task SafeExecute(Func<Task> action, string operation)
    {
        try
        {
            Console.WriteLine($"[LOG] {operation}");

            await action();
        }
        catch (Exception ex)
        {
            Console.WriteLine($"[CRASH] {operation}: {ex.Message}");
        }
    }

    private class PlayerInfo
    {
        public required int Id { get; set; }
        public required string Token { get; set; }
        public required string Name { get; set; }
    }

    private class GameData
    {
        public string CurrentEvent { get; set; } = string.Empty;

        public string Version { get; set; } = string.Empty;
    }

    private class PlayerData
    {
        public int Level { get; set; }

        public string Rank { get; set; } = string.Empty;



        public PlayerData()
        {
            Level = 1;

            Rank = "Default";
        }
    }

    private class RulesData
    {
        public string Mode { get; set; } = string.Empty;

        public string Map { get; set; } = string.Empty;
    }

    private class ActionData
    {
        public bool Ready { get; set; }
    }

    private class UpdateData
    {
        public int Round { get; set; }
        public string Message { get; set; } = string.Empty;
    }
}

File: java/Game.java
import com.fasterxml.jackson.annotation.JsonProperty;
import com.michitai.multiplayer.*;
import com.michitai.multiplayer.errors.*;
import com.michitai.multiplayer.games.*;
import com.michitai.multiplayer.leaderboard.*;
import com.michitai.multiplayer.matchmaking.*;
import com.michitai.multiplayer.matchmaking.requests.*;
import com.michitai.multiplayer.players.*;
import com.michitai.multiplayer.rooms.*;
import com.michitai.multiplayer.rooms.actions.*;
import com.michitai.multiplayer.rooms.realtime.*;
import com.michitai.multiplayer.rooms.updates.*;
import com.michitai.multiplayer.time.*;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;

public class Game {
    private static Client client;
    private static final Map<String, PlayerInfo> players = new HashMap<>();

    private static final String API_TOKEN = "YOUR_API_TOKEN";
    private static final String API_PRIVATE_TOKEN = "YOUR_PRIVATE_TOKEN";

    public static void main(String[] args) {
        System.out.println("=== MICHITAI Game SDK - Java Demo Started ===\n");

        Logger logger = new ConsoleLogger();
        client = new Client(
            API_TOKEN,
            API_PRIVATE_TOKEN,
            "https://api.michitai.com/api",
            logger
        );

        System.out.println("[INIT] SDK initialized with Jackson Databind");

        try {
            runAllDemos();
        } catch (Exception e) {
            System.err.println("[FATAL] Unexpected error: " + e.getMessage());
            e.printStackTrace();
        }

        System.out.println("=== All Demos Finished ===");
    }

    private static void runAllDemos() throws Exception {
        runDemoWithJoinByRequests();
        cleanupEverything();

        runDemoWithoutJoinByRequests();
        cleanupEverything();

        runDemoDirectRoom();
        cleanupEverything();

        runCommonTests();
    }

    // ====================== COMMON TESTS ======================
    private static void runCommonTests() throws Exception {
        System.out.println("\n=== COMMON TESTS: TIME, LEADERBOARD, GAME DATA ===\n");

        // Game Data
        safeExecute(() -> {
            GameDataResponse<GameData> gd = Games.getGameData(client, GameData.class);
            System.out.println("[GAME DATA] Retrieved. Game ID: " + gd.getGameId());
            GameData gameData = gd.getData();
            System.out.println("[GAME DATA] Event: " + gameData.getCurrentEvent() + ", Version: " + gameData.getVersion());
            
            GameData newGameData = new GameData();
            newGameData.setCurrentEvent("SpringFestival");
            newGameData.setVersion("1.2.3");
            Games.updateGameData(client, newGameData);
            System.out.println("[GAME DATA] Global data updated");
        }, "Game Data");

        // Server Time
        safeExecute(() -> {
            ServerTimeResponse time = Time.getServerTime(client);
            System.out.println("[TIME] Server UTC: " + time.getUtc());
        }, "GetServerTime");

        safeExecute(() -> {
            ServerTimeWithOffsetResponse timeOffset = Time.getServerTimeWithOffset(client, 3);
            System.out.println("[TIME] Server UTC+3: " + timeOffset.getUtc());
        }, "GetServerTimeWithOffset");

        // Leaderboard
        safeExecute(() -> {
            LeaderboardResponse<PlayerData> lb = Leaderboard.getLeaderboard(
                client, 
                new String[]{"level", "wins"}, 
                10, 
                PlayerData.class
            );
            System.out.println("[LEADERBOARD] Top " + lb.getLeaderboard().size() + " players");
            if (!lb.getLeaderboard().isEmpty()) {
                LeaderboardPlayer<PlayerData> top = lb.getLeaderboard().get(0);
                System.out.println("[LEADERBOARD] #1: " + top.getPlayerName() + ", Level: " + top.getPlayerData().getLevel() + " (Rank " + top.getRank() + ")");
            }
        }, "GetLeaderboard");
    }

    // ===================================================================
    // DEMO 1: MATCHMAKING WITH JOIN REQUESTS (Approval Flow)
    // ===================================================================
    private static void runDemoWithJoinByRequests() throws Exception {
        System.out.println("\n=== DEMO 1: MATCHMAKING WITH JOIN REQUESTS ===\n");
        setupPlayers();

        String matchmakingId = createMatchmakingLobby("DEMO 1 Matchmaking", true);

        String req1 = requestToJoinMatchmaking(players.get("p1").getToken(), matchmakingId, null);
        checkJoinRequestStatus(players.get("p1").getToken(), req1);
        approveJoinRequest(players.get("host").getToken(), req1);

        String req2 = requestToJoinMatchmaking(players.get("p2").getToken(), matchmakingId, new PlayerData());
        checkJoinRequestStatus(players.get("p2").getToken(), req2);

        getCurrentMatchmakingStatus();

        approveJoinRequest(players.get("host").getToken(), req2);

        getCurrentMatchmakingStatus();
        getMatchmakingPlayers();

        String roomId = startMatchmakingAndCreateRoom();
        runGameRoomFlow(roomId);
    }

    // ===================================================================
    // DEMO 2: MATCHMAKING DIRECT JOIN
    // ===================================================================
    private static void runDemoWithoutJoinByRequests() throws Exception {
        System.out.println("\n=== DEMO 2: MATCHMAKING DIRECT JOIN ===\n");
        setupPlayers();

        String matchmakingId = createMatchmakingLobby("DEMO 2 Matchmaking", false);

        for (PlayerInfo p : players.values()) {
            if (p.getToken().equals(players.get("host").getToken())) continue;
            joinMatchmakingDirectly(p.getToken(), matchmakingId, new PlayerData());
        }

        getCurrentMatchmakingStatus();
        getMatchmakingPlayers();

        String roomId = startMatchmakingAndCreateRoom();
        runGameRoomFlow(roomId);
    }

    // ===================================================================
    // DEMO 3: DIRECT ROOM (No Matchmaking)
    // ===================================================================
    private static void runDemoDirectRoom() throws Exception {
        System.out.println("\n=== DEMO 3: DIRECT ROOM CREATION ===\n");
        setupPlayers();

        RoomCreateResponse create = Rooms.createRoom(
            client,
            players.get("host").getToken(),
            "Direct Battle Arena",
            4,
            null,
            false,
            true,
            false,
            new PlayerData(),
            new RulesData()
        );
        String roomId = create.getRoomId();

        joinRoom(players.get("p1").getToken(), roomId);
        joinRoom(players.get("p2").getToken(), roomId);

        runGameRoomFlow(roomId);
    }

    // ====================== SETUP ======================
    private static void setupPlayers() throws Exception {
        System.out.println("[SETUP] Registering players...");

        PlayerRegisterResponse h = registerPlayer("GameHost", new PlayerData(15, 42, "gold"));
        PlayerRegisterResponse p1 = registerPlayer("PlayerOne", new PlayerData(12, 28, "silver"));
        PlayerRegisterResponse p2 = registerPlayer("PlayerTwo", new PlayerData(10, 15, "bronze"));

        players.put("host", new PlayerInfo(h.getPlayerId(), h.getPrivateKey(), "GameHost"));
        players.put("p1", new PlayerInfo(p1.getPlayerId(), p1.getPrivateKey(), "PlayerOne"));
        players.put("p2", new PlayerInfo(p2.getPlayerId(), p2.getPrivateKey(), "PlayerTwo"));

        // Authenticate all players
        for (PlayerInfo p : players.values()) {
            authenticatePlayer(p.getToken());
        }

        // Send players heartbeat
        for (PlayerInfo p : players.values()) {
            sendPlayerHeartbeat(p.getToken());
        }

        getAllPlayersList();

        for (PlayerInfo p : players.values()) {
            PlayerDataResponse<PlayerData> data = getPlayerData(p.getToken());
            PlayerData player = data.getData();
            player.setLevel(player.getLevel() + 1);
            updatePlayerData(p.getToken(), player);
            data = getPlayerData(p.getToken());
        }
    }

    private static void cleanupEverything() {
        System.out.println("\n[CLEANUP] Logging out players...");

        for (PlayerInfo p : players.values()) {
            safeExecute(() -> Players.logoutPlayer(client, p.getToken()), "Logout " + p.getName());
        }

        players.clear();
    }

    // ====================== HELPER METHODS ======================
    private static PlayerRegisterResponse registerPlayer(String name, PlayerData playerData) throws IOException {
        PlayerRegisterResponse reg = Players.registerPlayer(client, name, playerData);
        System.out.println("[REGISTER] " + name + " registered");
        return reg;
    }

    private static PlayerAuthResponse<PlayerData> authenticatePlayer(String token) throws IOException {
        PlayerAuthResponse<PlayerData> auth = Players.authenticatePlayer(client, token, PlayerData.class);
        System.out.println("[AUTH] " + auth.getPlayer().getPlayerName() + " authenticated");
        return auth;
    }

    private static PlayerHeartbeatResponse sendPlayerHeartbeat(String token) throws IOException {
        PlayerHeartbeatResponse heartbeat = Players.sendPlayerHeartbeat(client, token);
        System.out.println("[HEARTBEAT] Player heartbeat sent");
        return heartbeat;
    }

    private static void getAllPlayersList() throws IOException {
        PlayerListResponse list = Games.getAllPlayers(client);
        System.out.println("[PLAYERS LIST] Total: " + list.getCount());

        for (PlayerShort player : list.getPlayers()) {
            System.out.println("[PLAYERS LIST] Id: " + player.getId() + ", Name: " + player.getPlayerName() + 
                ", Online: " + player.isOnline() + ", Login: " + player.getLastLogin() + 
                ", Created: " + player.getCreatedAt());
        }
    }

    private static PlayerDataResponse<PlayerData> getPlayerData(String token) throws IOException {
        PlayerDataResponse<PlayerData> data = Players.getPlayerData(client, token, PlayerData.class);
        System.out.println("[PLAYER DATA] Player data retrieved");
        return data;
    }

    private static SuccessResponse updatePlayerData(String token, PlayerData data) throws IOException {
        SuccessResponse res = Players.updatePlayerData(client, token, data);
        System.out.println("[PLAYER DATA] Player data updated");
        return res;
    }

    private static String createMatchmakingLobby(String matchmakingName, boolean joinByRequests) throws IOException {
        RulesData rules = new RulesData();
        rules.setMode("tdm");
        rules.setMap("arena");

        PlayerData playerData = new PlayerData();
        playerData.setLevel(3);
        playerData.setWins(5);
        playerData.setRank("Diamond");

        MatchmakingCreateResponse res = Requests.createMatchmakingLobby(
            client,
            players.get("host").getToken(),
            matchmakingName,
            4,
            false,
            joinByRequests,
            false,
            false,
            false,
            null,
            playerData,
            rules
        );

        System.out.println("[MATCHMAKING] Lobby created (requests mode: " + joinByRequests + ")");
        return res.getMatchmakingId();
    }

    private static String requestToJoinMatchmaking(String token, String matchmakingId, PlayerData playerData) throws IOException {
        MatchmakingJoinRequestResponse req = Requests.requestToJoinMatchmaking(client, token, matchmakingId, playerData);
        System.out.println("[REQUEST] Sent: " + req.getRequestId());
        return req.getRequestId();
    }

    private static void checkJoinRequestStatus(String token, String requestId) throws IOException {
        MatchmakingRequestStatusResponse status = Requests.checkJoinRequestStatus(client, token, requestId);
        System.out.println("[REQUEST STATUS] " + status.getRequest().getStatus());
    }

    private static void approveJoinRequest(String hostToken, String requestId) throws IOException {
        MatchmakingPermissionResponse resp = Requests.respondToJoinRequest(
            client, 
            hostToken, 
            requestId, 
            EMatchmakingRequestAction.APPROVE
        );
        System.out.println("[APPROVE] " + resp.getMessage());
    }

    private static void joinMatchmakingDirectly(String token, String matchmakingId, PlayerData playerData) throws IOException {
        Matchmaking.joinMatchmakingDirectly(client, token, matchmakingId, playerData);
        System.out.println("[JOIN] Player joined matchmaking directly");
    }

    private static void getCurrentMatchmakingStatus() throws IOException {
        MatchmakingCurrentResponse<RulesData> s = Matchmaking.getCurrentMatchmakingStatus(
            client, 
            players.get("host").getToken(), 
            RulesData.class
        );

        System.out.println("[MATCHMAKING STATUS] Players in lobby: " + 
            (s.getMatchmaking() != null ? s.getMatchmaking().getCurrentPlayers() : 0));
    }

    private static void getMatchmakingPlayers() throws IOException {
        MatchmakingPlayersResponse<PlayerData> list = Matchmaking.getMatchmakingPlayers(
            client, 
            players.get("host").getToken(), 
            PlayerData.class
        );
        System.out.println("[MATCHMAKING PLAYERS] " + list.getPlayers().size() + " players");
    }

    private static String startMatchmakingAndCreateRoom() throws IOException {
        MatchmakingStartResponse start = Matchmaking.startGameFromMatchmaking(client, players.get("host").getToken());
        System.out.println("[START] Room created: " + start.getRoomId());
        return start.getRoomId();
    }

    private static void joinRoom(String token, String roomId) throws IOException {
        Rooms.joinRoom(client, token, roomId, null, new PlayerData());
        System.out.println("[ROOM] Player joined room " + roomId);
    }

    private static void runGameRoomFlow(String roomId) throws Exception {
        System.out.println("\n=== GAME ROOM FLOW ===\n");

        Rooms.getRooms(client, null, null, RulesData.class);

        CurrentRoomResponse<RulesData> room = Rooms.getCurrentRoom(
            client, 
            players.get("host").getToken(), 
            RulesData.class
        );

        // Players submit actions
        for (PlayerInfo p : players.values()) {
            safeExecute(() -> {
                ActionData actionData = new ActionData();
                actionData.setReady(true);
                
                SubmitAction<ActionData> actionReq = new SubmitAction<>(
                    ERoomTargetPlayers.HOST, 
                    "player_ready", 
                    actionData
                );
                Actions.submitAction(client, p.getToken(), actionReq);
            }, "SubmitAction " + p.getName());
        }

        // Host checks pending actions
        safeExecute(() -> {
            ActionPendingResponse<ActionData> pending = Actions.getPendingActions(
                client, 
                players.get("host").getToken(), 
                ActionData.class
            );
            System.out.println("[PENDING ACTIONS] " + pending.getActions().size() + " actions");
        }, "GetPendingActions");

        // Host broadcasts update
        safeExecute(() -> {
            UpdateData updateData = new UpdateData();
            updateData.setRound(1);
            updateData.setMessage("Game Started!");
            
            UpdatePlayers<UpdateData> updateReq = new UpdatePlayers<>(
                ERoomTargetPlayers.ALL, 
                "game_start", 
                updateData
            );
            Updates.updatePlayers(client, players.get("host").getToken(), updateReq);
            System.out.println("[UPDATE] Broadcast sent to all players");
        }, "Send Room Update");

        // Players poll updates
        for (PlayerInfo p : players.values()) {
            safeExecute(() -> {
                PollUpdates pollReq = new PollUpdates(ERoomTargetPlayers.HOST);
                Updates.pollUpdates(client, p.getToken(), pollReq);
            }, "PollUpdates " + p.getName());
        }

        Rooms.getRoomPlayers(client, players.get("host").getToken());

        // Heartbeats
        for (PlayerInfo p : players.values()) {
            safeExecute(() -> Rooms.sendRoomHeartbeat(client, p.getToken()), "RoomHeartbeat " + p.getName());
        }

        // Stop room
        safeExecute(() -> Rooms.stopRoom(client, players.get("host").getToken()), "StopRoom " + room.getRoom().getRoomName());
    }

    private static void safeExecute(ThrowingRunnable action, String operation) {
        try {
            System.out.println("[LOG] " + operation);
            action.run();
        } catch (Exception e) {
            System.err.println("[CRASH] " + operation + ": " + e.getMessage());
            e.printStackTrace();
        }
    }

    @FunctionalInterface
    private interface ThrowingRunnable {
        void run() throws Exception;
    }

    // ====================== PLAYER INFO ======================
    private static class PlayerInfo {
        private int id;
        private String token;
        private String name;

        public PlayerInfo(int id, String token, String name) {
            this.id = id;
            this.token = token;
            this.name = name;
        }

        public int getId() { return id; }
        public String getToken() { return token; }
        public String getName() { return name; }
    }

    // ====================== GAME DATA ========================
    private static class GameData {
        @JsonProperty("currentEvent")
        private String currentEvent;
        
        @JsonProperty("version")
        private String version;

        public String getCurrentEvent() { return currentEvent; }
        public void setCurrentEvent(String currentEvent) { this.currentEvent = currentEvent; }
        
        public String getVersion() { return version; }
        public void setVersion(String version) { this.version = version; }
    }

    // ====================== PLAYER DATA ======================
    private static class PlayerData {
        @JsonProperty("level")
        private int level;
        
        @JsonProperty("wins")
        private int wins;
        
        @JsonProperty("rank")
        private String rank;

        public PlayerData() {
            this.level = 1;
            this.wins = 0;
            this.rank = "Default";
        }

        public PlayerData(int level, int wins, String rank) {
            this.level = level;
            this.wins = wins;
            this.rank = rank;
        }

        public int getLevel() { return level; }
        public void setLevel(int level) { this.level = level; }
        
        public int getWins() { return wins; }
        public void setWins(int wins) { this.wins = wins; }
        
        public String getRank() { return rank; }
        public void setRank(String rank) { this.rank = rank; }
    }

    // ====================== MATCHMAKING DATA =================
    private static class RulesData {
        @JsonProperty("mode")
        private String mode;
        
        @JsonProperty("map")
        private String map;

        public String getMode() { return mode; }
        public void setMode(String mode) { this.mode = mode; }
        
        public String getMap() { return map; }
        public void setMap(String map) { this.map = map; }
    }

    // ====================== ROOM DATA ========================
    private static class ActionData {
        @JsonProperty("ready")
        private boolean ready;

        public boolean isReady() { return ready; }
        public void setReady(boolean ready) { this.ready = ready; }
    }

    private static class UpdateData {
        @JsonProperty("round")
        private int round;
        
        @JsonProperty("message")
        private String message;

        public int getRound() { return round; }
        public void setRound(int round) { this.round = round; }
        
        public String getMessage() { return message; }
        public void setMessage(String message) { this.message = message; }
    }
}
